Update the device.app.lifecycle event description and constraints (#794)
open-telemetry/semantic-conventions: docs/mobile/events.md

`os.name` [resource semantic convention](/docs/resource/os.md) to identify the
mobile operating system (e.g. Android, iOS).

The following table describes the payload fields that MUST
be used to describe the state of the application at the time of the event.

The `android.state` and `ios.state` fields are mutually exclusive and MUST
NOT be used together, each field MUST be used with its corresponding
`os.name` [resource semantic convention](/docs/resource/os.md) value.

### Event details

<!-- semconv device.app.lifecycle(full) -->
The event name MUST be `device.app.lifecycle`.

> 

Reviewer: Can we document the severity level for this event since it's part of Event API and important for filtering?

Reviewer 2: Severity is not in the payload though, and instrumentations can log the event at any severity level.

Author: There is no severity associated with the event, and severity is an Optional field and is not required for this event.

`os.name` [resource semantic convention](/docs/resource/os.md) to identify the
mobile operating system (e.g. Android, iOS).

> Payload field MUST be used to describe the state of the application at the
> time of the event. The following table describes the payload fields that MUST
> be used to describe the state of the application at the time of the event.

Reviewer: ```suggestion
The following table describes the payload fields that MUST
be used to describe the state of the application at the time of the event.
```

The `android.state` and `ios.state` fields are mutually exclusive and MUST
NOT be used together, each field MUST be used with its corresponding

> `os.name` [resource semantic convention](/docs/resource/os.md) value.

Reviewer: Is it still needed?

Author: I'm thinking of leaving this for now (as part of this PR) as it provides a link to the OS semantic conventions and highlights it a bit more than just the individual requirement level (at least for now).

Once we have the build tools defined and available this can be reviewed again then.

`os.name` [resource semantic convention](/docs/resource/os.md) to identify the
mobile operating system (e.g. Android, iOS).

> Payload attributes MUST be used to describe the state of the application at the

Reviewer: is "payload attribute" a spec terminology? If not, can we call them "event fields" to avoid confusion ?

Author: The `payload` is what we have been calling main payload of an event, the wording here (as I'm just reformatting) pre-dates calling these `payload fields`, I'll update

time of the event. The following table describes the payload attributes that MUST
be used to describe the state of the application at the time of the event.

The `android.state` and `ios.state` attributes are mutually exclusive and MUST

> NOT be used together, each attribute MUST be used with it's corresponding

Reviewer 3: ```suggestion
NOT be used together, each attribute MUST be used with its corresponding
```

`os.name` [resource semantic convention](/docs/resource/os.md) value. For
example, if the `os.name` attribute is set to `android` then the
`android.state` attribute MUST be used and the `ios.state` attribute MUST NOT
be used. If the `os.name` attribute is set to `ios` then the `ios.state`

> attribute MUST be used and the `android.state` attribute MUST NOT be used.

Reviewer 3: I don't think the "For example...." part makes it any clearer (a little rambly to me). I think just saying that exactly one of `ios.state` or `android.state` must be present.

Reviewer 4: I'm also wondering whether we need 2 separate "*.state" attrs each prefixed with the same value that is also present in the attr `os.name`.

Author: I'm going to leave the 2 states for this PR and that can be a follow up one, I have removed for example as @lmolkova suggestion of the requirement level I think makes it clearer.

### Event details

<!-- semconv device.app.lifecycle(full) -->

> The event name MUST be `device.app.lifecycle`.

Reviewer: not related to this PR, but it'd be nice to render a table here (see https://github.com/open-telemetry/build-tools/issues/292)

<!-- endsemconv -->

<!-- Manually adding the markdown table until the body definition is available in the build tools -->
| Body Field | Type | Description  | Examples  | Requirement Level |
|---|---|---|---|---|

> | `android.state` | string | This attribute represents the state the application has transitioned into at the occurrence of the event. [1] | `created` | See below |

Reviewer: I think using `constraints` makes markdown worse than is could be.  With `requirement_level: conditionally_required: if and only if `os.name` is `android`` it would look better and would eliminate the need for mutual exclusivity notes above and below.

```suggestion
| `android.state` | string | This attribute represents the state the application has transitioned into at the occurrence of the event. [1] | `created` | `Conditionally Required`: if and only if `os.name` is `android`|
```
